fishtest/utils/delta_update_users.py

```diff
@@ -0,0 +1,165 @@
+
+def build_users(machines, info):
+  for machine in machines:
+    games_per_hour = (machine['nps'] / 1200000.0) * (3600.0 / parse_tc(machine['run']['args']['tc'])) * int(machine['concurrency'])
+    info[machine['username']]['games_per_hour'] += games_per_hour
+
+  users = []
+  for u in info.keys():
+    user = info[u]
+    try:
+      user['last_updated'] = delta_date(user['last_updated'])
+    except:
+      pass
+    users.append(user)
+
+  users = [u for u in users if u['games'] > 0 or u['tests'] > 0]
+  return users
+
+def update_users():
+  rundb = RunDb()
+
+  info = {}
+  top_month = {}
+
+  last_stats = datetime.min
+  clear_stats = True
+  if len(sys.argv) > 1:
+    print('scan all')
+  else:
+    for stat in rundb.actiondb.get_actions(1, 'update_stats'):
+      last_stats = stat['time']
+      clear_stats = False
+  print('last: ' + str(last_stats))
+
+  for u in rundb.userdb.get_users():
+    username = u['username']
+    top_month[username] = {'username': username,
+                      'cpu_hours': 0,
+                      'games': 0,
+                      'tests': 0,
+                      'tests_repo': u.get('tests_repo', ''),
+                      'last_updated': datetime.min,
+                      'games_per_hour': 0.0,}
+    if clear_stats:
+      info[username] = top_month[username].copy()
+    else:
+      info[username] = rundb.userdb.user_cache.find_one({'username': username})
+      if not info[username]:
+        info[username] = top_month[username].copy()
+
+  for run in rundb.get_unfinished_runs():
+    process_run(run, top_month)
+
+  # Step through these in small batches (step size 100) to save RAM and
+  # minimize time in critical section (start with step size 1)
+  current = 0
+  step_size = 1
+  now = datetime.utcnow()
+
+  # prevent race condition with 'stop_run' in server
+  fcntl.flock(RunDb.lock_file, fcntl.LOCK_EX)
+  # record this update run
+  rundb.actiondb.update_stats()
+
+  more_days = True
+  first = True
+  while more_days:
+    runs = rundb.get_finished_runs(skip=current, limit=step_size)[0]
+    if first:
+      fcntl.flock(RunDb.lock_file, fcntl.LOCK_UN)
+      first = False
+      print('race window: ' + str(datetime.utcnow() - now))
+    if len(runs) == 0:
+      break
+    for run in runs:
+      process_run(run, info, last_stats)
+      if (now - run['start_time']).days < 31:
+        process_run(run, top_month)
+      elif not clear_stats:
+        more_days = False
+    current += step_size
+    step_size = 100
+
+  machines = rundb.get_machines()
+
+  users = build_users(machines, info)
+  rundb.userdb.user_cache.remove()
+  rundb.userdb.user_cache.insert(users)
+
+  rundb.userdb.top_month.remove()
+  rundb.userdb.top_month.insert(build_users(machines, top_month))
+
+  # Delete users that have never been active and old admins group
+  idle = {}
+  for u in rundb.userdb.get_users():
+      update = False
+      while 'group:admins' in u['groups']:
```

Comment:
If I'm not wrong this should be useful only once or after loading an old backup.

Reply:
Yes you are correct. Note that I intent to replace the locking (because it takes too long and does still not make the statistics accurate) with a small mongodb collection which stores the deltas of the last run. I will do that the next few days...

Reply:
Often I test monkey style w/o reading the code, so please give me the input after adding the new collection: I will update my mongodb maintenance script.

Reply:
You should not need to edit your scripts. Just run `delta_update_users.py` as you did before.
The first run should be slow, the later runs should be fast...